mono/utils/mono-log-windows.c

```diff
@@ -0,0 +1,129 @@
+/*
+ * mono-log-windows.c: Simplistic simulation of a syslog logger for Windows
+ *
+ * This module contains the Windows syslog logger interface
+ *
+ * Author:
+ *    Neale Ferguson <neale@sinenomine.net>
+ *
+ */
+#include <config.h>
+
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+
+#ifdef HOST_WIN32
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <ctype.h>
+#include <string.h>
+#include <glib.h>
+#include <errno.h>
+#include <time.h>
+#include "mono-logger.h"
+
+static FILE *logFile = NULL;
+static void *logUserData = NULL;
+static char *logFileName = L".//mono.log";
+
+/**
+ * mapSyslogLevel:
+ * 	
+ * 	@level - GLogLevelFlags value
+ * 	@returns The equivalent character identifier
+ */
+static inline char 
+mapLogFileLevel(GLogLevelFlags level) 
+{
+	if (level & G_LOG_LEVEL_ERROR)
+		return ('E');
+	if (level & G_LOG_LEVEL_CRITICAL)
+		return ('C');
+	if (level & G_LOG_LEVEL_WARNING)
+		return ('W');
+	if (level & G_LOG_LEVEL_MESSAGE)
+		return ('N');
+	if (level & G_LOG_LEVEL_INFO)
+		return ('I');
+	if (level & G_LOG_LEVEL_DEBUG)
+		return ('D');
+	return ('I');
+}
+
+/**
+ * mono_log_open_logfile
+ * 	
+ *	Open the logfile. If the path is not specified default to stdout. If the
+ *	open fails issue a warning and use stdout as the log file destination.
+ *
+ * 	@ident - Identifier: ignored
+ * 	@userData - Not used
+ */
+void
+mono_log_open_syslog(const char *ident, void *userData)
+{
+	logFile = fopen(logFileName, "w");
+	if (logFile == NULL) {
+		g_warning("opening of log file %s failed with %s",
+			  strerror(errno));
+	}
+	logUserData = userData;
+}
+
+/**
+ * mono_log_write_logfile
+ * 	
+ * 	Write data to the log file.
+ *
+ * 	@domain - Identifier string
+ * 	@level - Logging level flags
+ * 	@format - Printf format string
+ * 	@vargs - Variable argument list
+ */
+void
+mono_log_write_syslog(const char *domain, GLogLevelFlags level, mono_bool hdr, const char *format, va_list args)
+{
+	time_t t;
+	struct tm *tod;
+	char logTime[80],
+	      logMessage[512];
+	pid_t pid;
+	int iLog = 0;
+	size_t nLog;
+
+	if (logFile == NULL)
+		mono_log_open_logfile(NULL, NULL);
+
+	time(&t);
+	tod = localtime(&t);
+	pid = _getpid();
+	strftime(logTime, sizeof(logTime), "%F %T", tod);
+	iLog = snprintf(logMessage, sizeof(logMessage), "%s level[%c] mono[%d]: ",
+			logTime,mapLogFileLevel(level),pid);
+	nLog = sizeof(logMessage) - iLog - 2;
+	iLog = vsnprintf(logMessage+iLog, nLog, format, args);
+	logMessage[iLog++] = '\n';
+	logMessage[iLog++] = 0;
+	fputs(logMessage, logFile);
+	fflush(logFile);
+
+	if (level == G_LOG_FLAG_FATAL)
+		abort();
+}
```

Comment:
It is identical to common. This is a kludge until I can work out how to use the Windows Event logs or similar.

From: Johan Lorensson +ADw-notifications+AEA-github.com+ADw-mailto:notifications+AEA-github.com+AD4APg-
Reply-To: mono/mono +ADw-reply++0004b6fa29fb914ea6f7388bf09146248e88a3298129869892cf000000011369077792a170ce03ea20AQA-reply.github.com+ADw-mailto:reply++0004b6fa29fb914ea6f7388bf09146248e88a3298129869892cf000000011369077792a170ce03ea20AQA-reply.github.com+AD4APg-
Date: Friday, June 3, 2016 at 4:53 AM
To: mono/mono +ADw-mono+AEA-noreply.github.com+ADw-mailto:mono+AEA-noreply.github.com+AD4APg-
Cc: Neale Ferguson +ADw-neale+AEA-sinenomine.net+ADw-mailto:neale+AEA-sinenomine.net+AD4APg-, Mention +ADw-mention+AEA-noreply.github.com+ADw-mailto:mention+AEA-noreply.github.com+AD4APg-
Subject: Re: +AFs-mono/mono+AF0- Enhance log tracing (+ACM-2734)


In mono/utils/mono-log-windows.c+ADw-https://github.com/mono/mono/pull/2734+ACM-discussion+AF8-r65675674+AD4-:

+AD4-       time(+ACY-t)+ADs-
+AD4-       tod +AD0- localtime(+ACY-t)+ADs-
+AD4-       pid +AD0- +AF8-getpid()+ADs-
+AD4-       strftime(logTime, sizeof(logTime), +ACIAJQ-F +ACU-T+ACI-, tod)+ADs-
+AD4-       iLog +AD0- snprintf(logMessage, sizeof(logMessage), +ACIAJQ-s level+AFsAJQ-c+AF0- mono+AFsAJQ-d+AF0-: +ACI-,
+AD4-                       logTime,mapLogFileLevel(level),pid)+ADs-
+AD4-       nLog +AD0- sizeof(logMessage) - iLog - 2+ADs-
+AD4-       iLog +AD0- vsnprintf(logMessage++iI-, nLog, format, args)+ADs-
+AD4-       logMessage+AFs-iLog+++8AXQ- +AD0- '+AFw-n'+ADs-
+AD4-       logMessage+AFs-iLog+++8AXQ- +AD0- 0+ADs-
+AD4-       fputs(logMessage, logFile)+ADs-
+AD4-       fflush(logFile)+ADs-
+AD4-
+AD4-       if (level +AD0APQ- G+AF8-LOG+AF8-FLAG+AF8-FATAL)
+AD4-               abort()+ADs-
+AD4- +AH0-


Possible to share this code with common, seems like duplication of most of the code in mono+AF8-log+AF8-write+AF8-logfile?

+IBQ-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub+ADw-https://github.com/mono/mono/pull/2734/files/583d1a2580a5fc61d0002920c17c23fa71375c3d+ACM-r65675674+AD4-, or mute the thread+ADw-https://github.com/notifications/unsubscribe/AAS2-qk2RTy2eblXZI75t2o46UDQy+AF8-0uks5qH-t3gaJpZM4HsGtY+AD4-.